test/e2e/network/topology_hints.go

```diff
@@ -153,7 +153,7 @@ var _ = common.SIGDescribe("Feature:Topology Hints", func() {
 				framework.Failf("Expected zone to be specified for %s node", nodeName)
 			}
 			ginkgo.By("creating a client pod for probing the service from " + fromZone)
-			podName := "curl-from-" + fromZone
+			podName := fmt.Sprintf("client%d-curl-from-%s", i, fromZone)
```

Comment:
I think that we leaked the environment assumptions into the test, this is an example of an execution.
```
[BeforeEach] [sig-network] Feature:Topology Hints
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/topology_hints.go:47
[It] should distribute endpoints evenly
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/topology_hints.go:52
Jan 14 06:43:17.801: INFO: DaemonSet pods can't tolerate node kind-control-plane with taints [{Key:node-role.kubernetes.io/master Value: Effect:NoSchedule TimeAdded:<nil>}], skip checking this node
Jan 14 06:43:17.808: INFO: Number of nodes with available pods controlled by daemonset topology-serve-hostname: 0
Jan 14 06:43:17.808: INFO: Node kind-worker is running 0 daemon pod, expected 1
Jan 14 06:43:22.799: INFO: DaemonSet pods can't tolerate node kind-control-plane with taints [{Key:node-role.kubernetes.io/master Value: Effect:NoSchedule TimeAdded:<nil>}], skip checking this node
Jan 14 06:43:22.803: INFO: Number of nodes with available pods controlled by daemonset topology-serve-hostname: 3
Jan 14 06:43:22.803: INFO: Number of running nodes: 3, number of available pods: 3 in daemonset topology-serve-hostname
Jan 14 06:43:22.808: INFO: DaemonSet pods can't tolerate node kind-control-plane with taints [{Key:node-role.kubernetes.io/master Value: Effect:NoSchedule TimeAdded:<nil>}], skip checking this node
Jan 14 06:43:22.808: INFO: Waiting for 3 endpoints to be tracked in EndpointSlices
[1mSTEP[0m: having hints set for each endpoint
[1mSTEP[0m: keeping requests in the same zone
[1mSTEP[0m: creating a client pod for probing the service from zone-a
Jan 14 06:43:27.830: INFO: The status of Pod curl-from-zone-a is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:29.834: INFO: The status of Pod curl-from-zone-a is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:31.835: INFO: The status of Pod curl-from-zone-a is Running (Ready = true)
Jan 14 06:43:31.837: INFO: Ensuring that requests from curl-from-zone-a pod on kind-worker node stay in zone-a zone
Jan 14 06:43:36.850: INFO: Pod client logs: Fri Jan 14 06:43:29 UTC 2022
Date: Fri Jan 14 06:43:30 UTC 2022 Try: 1
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:31 UTC 2022 Try: 2
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:32 UTC 2022 Try: 3
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:33 UTC 2022 Try: 4
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:34 UTC 2022 Try: 5
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:35 UTC 2022 Try: 6
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:36 UTC 2022 Try: 7
topology-serve-hostname-cvlkk

[1mSTEP[0m: creating a client pod for probing the service from zone-b
Jan 14 06:43:36.858: INFO: The status of Pod curl-from-zone-b is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:38.866: INFO: The status of Pod curl-from-zone-b is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:40.863: INFO: The status of Pod curl-from-zone-b is Running (Ready = true)
Jan 14 06:43:40.866: INFO: Ensuring that requests from curl-from-zone-b pod on kind-worker2 node stay in zone-b zone
Jan 14 06:43:45.873: INFO: Pod client logs: Fri Jan 14 06:43:38 UTC 2022
Date: Fri Jan 14 06:43:39 UTC 2022 Try: 1
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:40 UTC 2022 Try: 2
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:41 UTC 2022 Try: 3
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:42 UTC 2022 Try: 4
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:43 UTC 2022 Try: 5
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:44 UTC 2022 Try: 6
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:45 UTC 2022 Try: 7
topology-serve-hostname-fx8kl

[1mSTEP[0m: creating a client pod for probing the service from zone-c
Jan 14 06:43:45.882: INFO: The status of Pod curl-from-zone-c is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:47.888: INFO: The status of Pod curl-from-zone-c is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:49.886: INFO: The status of Pod curl-from-zone-c is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:51.887: INFO: The status of Pod curl-from-zone-c is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:53.886: INFO: The status of Pod curl-from-zone-c is Running (Ready = true)
Jan 14 06:43:53.892: INFO: Ensuring that requests from curl-from-zone-c pod on kind-worker3 node stay in zone-c zone
Jan 14 06:43:58.900: INFO: Pod client logs: Fri Jan 14 06:43:46 UTC 2022
Date: Fri Jan 14 06:43:47 UTC 2022 Try: 1
````

If I recall correctly this loop has to iterate for one node of each zone, so it checks that for all existing zones in the cluster, the traffic remains in the zone.

Based on that, the name is correct, what is not correct is the loop, that instead of iterating over the nodeNames and break at 3, it has to iterate over the zones and pick one node of each zone, this will guarantee that the pods names are unique too

Let's wait for @robscott for confirmation

Reply:
Yep, this test assumes it's in a cluster with a 3 nodes, each in different zones. It would make sense to update this test to support more flexible deployments and bail out/error if it is in an unsupported one.

For reference, this test currently assumes the following:
1. All nodes have equivalent allocatable CPU
2. There are 3 nodes, each in a different zone
3. The TopologyAwareHints feature gate is enabled

I don't think we have a way to verify 3, but we could add a check for 1, and refactor the test to work if the number of nodes/zones is different. We still would need nodes to exist in more than one zone for this test to work though.

Reply:
@aojea @robscott thank you for your suggestionsi and I have updated the code. PTAL.

Reply:
don't add new helper functions, until we have clear we can't reuse them I prefer to keep that code within the test.
Another thing is that you can do a List instead of a Get per node, it is going to be more efficiente and reduce flakiness

Reply:
Makes sense. Code is updated.